### PR TITLE
typing for numpy 1.21

### LINKED ISSUE #5517: mypy issues on `main`
The `pre-commit` is reporting some `mypy` issues on the `main` branch:
```
xarray/core/formatting.py:193: error: No overload variant of "max" matches argument types "signedinteger[Any]", "int"  [call-overload]
xarray/core/formatting.py:193: note: Possible overload variant:
xarray/core/formatting.py:193: note:     def [SupportsLessThanT <: SupportsLessThan] max(__arg1, SupportsLessThanT, SupportsLessThanT, *_args: SupportsLessThanT, key: None = ...) -> SupportsLessThanT
xarray/core/formatting.py:193: note:     <5 more non-matching overloads not shown>
xarray/core/indexes.py:166: error: Incompatible types in assignment (expression has type "Union[dtype[object_], dtype[Any], dtype[void]]", variable has type "dtype[object_]")  [assignment]
xarray/core/computation.py:607: error: Argument 1 to "append" of "list" has incompatible type "None"; expected "slice"  [arg-type]
xarray/core/accessor_str.py:277: error: <nothing> not callable  [misc]
Found 4 errors in 4 files (checked 143 source files)
```

does anyone know how to fix those?

### PR DESCRIPTION
**Pull request by mathause**
- [x] Closes  #5517
- [x] Passes `pre-commit run --all-files`

The minimal typing for numpy 1.21. As always I am by no means a typing specialist.

### CODE REVIEW

File: xarray/core/computation.py [hunk 1]
```diff
@@ -597,7 +597,7 @@ def broadcast_compat_data(
         data = duck_array_ops.transpose(data, order)
 
     if new_dims != reordered_dims:
-        key_parts = []
+        key_parts: list[Union[slice, None]] = []
```
Comment (max-sixty) [inline xarray/core/computation.py:600]: FYI this list is python3.9+, probably fine but maybe safer to use `List`?

```suggestion
        key_parts: List[Optional[slice]] = []
```

**Review by max-sixty — APPROVED** [review]
Thanks @mathause !

File: xarray/core/computation.py (hunk 1 above)
Comment (mathause) [reply xarray/core/computation.py]: > FYI this list is python3.9+

Ah right, yes let's use `List`

**Comment by keewis** [comment]
thanks, @mathause. Given that this gets `pre-commit` to pass I'll merge this now.

**Pull request merged**